packages/@jsii/java-runtime-test/project/src/test/java/software/amazon/jsii/testing/TypeCheckingTest.java

```diff
@@ -0,0 +1,124 @@
+package software.amazon.jsii.testing;
+
+import org.junit.jupiter.api.Test;
+import software.amazon.jsii.JsiiException;
+import software.amazon.jsii.JsiiObject;
+import software.amazon.jsii.tests.calculator.*;
+import software.amazon.jsii.tests.calculator.anonymous.*;
+import software.amazon.jsii.tests.calculator.anonymous.UseOptions;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class StructAImplementer implements StructA, StructB {
+  public String requiredString;
+
+  StructAImplementer(String param) {
+    requiredString = param;
+  }
+
+  public void setRequiredString(String param) {
+    requiredString = param;
+  }
+
+  public String getRequiredString() {
+    return requiredString;
+  }
+}
+
+public class TypeCheckingTest {
+    @Test
+    public void constructor() {
+        HashMap<String, Object> map = new HashMap<String, Object>();
+        map.put("good", new StructAImplementer("present"));
+        map.put("bad", "Not a StructA or StructB");
+        ArrayList<Map<String, Object>> list = new ArrayList<Map<String, Object>>();
+        list.add(map);
+        Exception e = assertThrows(IllegalArgumentException.class, () ->
+        {
+            new ClassWithCollectionOfUnions(list);
+        });
+
+        assertEquals("Expected unionProperty.get(0).get(\"bad\") to be one of: software.amazon.jsii.tests.calculator.StructA, software.amazon.jsii.tests.calculator.StructB; received class java.lang.String, (Parameter 'unionProperty')", e.getMessage());
```

Comment:
The "(Parameter 'unionProperty')" part here is a .NET-ism (the .NET exception records the parameter name and ALWAYS appends it to the error message like so)... I don't mind it being present, though... But the parameter name is actually already part of the error message.